Feature/robot page objects (#1048)
SFDO-Tooling/CumulusCI: docs/robotframework.rst

* **Salesforce Update**: Updates a record using its type, ID, and field=value syntax
* **SOQL Query**: Runs a SOQL query and returns a REST API result dictionary

PageObjects Library
===================

The **PageObjects** library provides support for page objects,

> robotframework-style. Even though robot is a keyword driven framework,

Reviewer: ```suggestion
Robot Framework-style. Even though robot is a keyword-driven framework,
```

we've implemented a way to dynamically load in keywords that are
unique to a page or an object on the page.

With this library, you can define classes which represent page
objects. Each class has keywords that are unique to a page or a
component. These classes can be imported on demand only for tests
which use these pages or components.

> The pageobject Decorator

Reviewer: ```suggestion
The ``pageobject`` Decorator
```

------------------------

> Page objects are normal python classes which use the :code:`pageobject`

Reviewer: ```suggestion
Page objects are normal Python classes which use the :code:`pageobject`
```

> decorator provided by cumulusci. Unlike traditional robot framework

Reviewer: ```suggestion
decorator provided by CumulusCI. Unlike traditional Robot Framework
```

keyword libraries, you may define multiple sets of keywords in a
single file.

For example, the following example shows the definition of a page
object for the listing page of the Contact object. It adds a new
keyword named :code:`Click contact link`:

.. code-block:: python

   from cumulusci.robotframework.pageobjects import pageobject, ListingPage

   @pageobject(page_type="Listing", object_name="Contact")
   class ContactListingPage(ListingPage):

       def click_contact_link(self, name):
           self.selenium.click_link('xpath://a[@title="{}"]'.format(name))
           self.salesforce.wait_until_loading_is_complete()

The :code:`pageobject` decorator takes two arguments: :code:`page_type` and
:code:`object_name`. These two arguments are used to identify the page
object (eg: :code:`Go To Page  Listing  Contact`). The values can be
any arbitrary string, but ordinarily should represent standard page

> types ("Listing", "Detail", "Home"), and standard object names.

Reviewer: The developer name or the user-facing label for the object? If the former, we need to talk about how to make this work both when objects have a namespace prefix and when they don't.

Author: Yes, I haven't given namespaces any consideration. I haven't yet taken the time to grok how they work.

Reviewer: It's not super complicated (though dealing with the implications can be), and I think it's worth grokking before we introduce a framework that assumes objects are always named the same thing.

Standard objects like Contact always have the same name.

Custom objects which are not in a managed package or a namespaced scratch org have a name with a `__c` suffix like MyObject__c. So far no problem.

Custom objects which are in a managed package add a prefix using the package's globally unique namespace. For example npsp__Address__c is the developer name of the Address object from NPSP when it is installed as a managed package. Objects also get this namespace prefix if they are in an unmanaged package being developed in a scratch org that has its own namespace.

Since our products are typically developed as unmanaged packages and then released as managed packages, that means that the Address object would be called Address__c during development but npsp__Address__c once it is released (or in development in a namespaced scratch org).

Since we want to run our robot tests both during development and with managed packages for regression, we need to figure out how to handle this.

Author: We probably need to discuss this over chat or a video call, but let me get a couple of basic questions out of the way.

At the time that a developer needs to write a keyword, do they know what the namespace is / will be? Could the solution be as simple as defining the namespace in the class or decorator, and then at runtime we do something like this:

    @pageobject(page_type="Listing", object_name="MyObject__c", namespace="npsp")
    def whatever(ListingPage):
        ...
        object_name = self.namespace + self.object_name if self.is_managed() else self.object_name
        url = ".../lightning/o/{}/list".format(object_name)"
        ...

Or is it a bit more complicated than that? Is it possible to write a `self.is_managed` function or property that can use introspection to get the answer (and/or a `self.get_namespace_prefix()` function)?

Author: poking about through the NPSP code I see that it does `self.cumulusci.sf.describe()` and then does a bit of string manipulation to get the namespace prefix. Would that be the foundation of `self.is_managed()` or `self.get_namespace_prefix()`?

Reviewer: That sounds like the right direction, but we might not even need to add the `namespace` argument because it's typically only objects from the current repository that would vary between namespaced and un-namespaced, so we can look up the current project's namespace from cumulusci.yml. (Underlying packages would be installed as managed packages and always have their namespace.)

There's already a keyword in the CumulusCI library, `Get Namespace Prefix` to get the namespace prefix for a specified package (defaulting to the package for the current repository) or an empty string if there's no prefix.

`object_name` could become a property that encapsulates the logic for adding or not-adding the prefix (and memoizes it).

Author: Will namespaces always only apply to custom objects? That is, if I set `object_name` to `Contact`, I can determine it will never have a namespace associated with it, but if the object is named `MyObject__c`, then I need to do the dance to maybe prepend the namespace?

I (think I) want to support the ability for a teams to create page objects for standard objects, until we reach the point that we have full coverage. For example, NPSP or GEM or whomever might create a `Contact` page object or `Opportunity` page object to get them over a hump, which we can eventually fold into the core. In such a case, we wouldn't want to use the namespace logic for those page objects.

So, the logic for building a URL might look like:

```
    if object_name.endswith("__c") and is_managed():
        prepend the namespace
    else
        don't prepend the namespace
```

Reviewer: That's right that it's only for custom objects. It's also only for custom objects that don't already have a different namespace. I think the logic could be something like this:

    @property
    def object_name(self):
        object_name = self._object_name # assuming we change the decorator to set _object_name instead of object_name
        parts = object_name.split("__")
        if len(parts) == 2 and parts[-1] == "c":  # the length check is to skip objects from a different namespace like foobar__otherpackageobject__c
            object_name = self.cumulusci.get_namespace_prefix() + object_name # get_namespace_prefix already takes care of returning an actual prefix or an empty string depending on whether the package is managed
        return object_name

Importing the library
---------------------

The **PageObjects** library is somewhat unique in that it is not only a
keyword library, but also the mechanism by which you can import files
which contain page object classes. This is done by providing the paths

> to one or more python files which implement page objects. You may also

Reviewer: ```suggestion
to one or more Python files which implement page objects. You may also
```

import **PageObjects** without passing any files to it in order to take
advantage of some general purpose page objects.

For example, consider the case where you've created two files that
each have one or more page object definitions. For example, lets say

> in robot/MyProject/resources you have the files PageObjects.py and

Reviewer: ```suggestion
in ``robot/MyProject/resources`` you have the files ``PageObjects.py`` and
```

> MorePageObjects.py. You can import these page objects into a test

Reviewer: ```suggestion
``MorePageObjects.py``. You can import these page objects into a test
```

represents the current page. Each page object may use its own method
for making the determination, but the built-in page objects all
compare the page location to an expected pattern
(eg: _.../lightning/o/..._). If the assertion passes, the keywords for
that page object will autoamtically be loaded.

This keyword is useful if you get to a page via a button or some other
form of navigation, in that it allows you to both assert that you are
on the page you think you should be on, and load the keywords for that
page, all with a single statement.

Go To Page Object
^^^^^^^^^^^^^^^^^

Example: :code:`Go to page object  Listing  Contact`

This will attempt to go to the listing page for the Contact object,
and then load the keywords for that page.

Log Page Object Keywords
^^^^^^^^^^^^^^^^^^^^^^^^

Example: :code:`Load Page Object Keywords`

This keyword is primarily a debugging tool. When called it will log
each of the keywords for the current page object.

Load Page Object
^^^^^^^^^^^^^^^^

Example: :code:`Load page object  Listing  Contact`

This will load the page object for the given **page_type** and
**object_name_**. It is useful when you want to use the keywords from a

> page object without first navigating to that page (ie: when you are

Reviewer: ```suggestion
page object without first navigating to that page (i.e. when you are
```

already on the page and don't want to navigate away).

Generic Page Objects
--------------------

You do not need to create a page object in order to take advantage of
the new page object keywords. If you use one of the page object
keywords for a page that does not have its own page object, the
**PageObjects** library will try to find a generic page.

For example, if you use :code:`Current page should be  Home  Event` and
there is no page object by that name, a generic :code:`Home` page object

> will be loaded, and it's object name will be set to :code:`Event`.

Reviewer: ```suggestion
will be loaded, and its object name will be set to :code:`Event`.
```

Author: ugh. I hate it when I make this mistake.

> For example, lets say your project has created a custom object named

Reviewer: ```suggestion
For example, let's say your project has created a custom object named
```
